src/JustSaying/Extensions/ChannelExtensions.cs

```diff
@@ -9,39 +10,49 @@ namespace JustSaying.Extensions
 {
     public static class ChannelExtensions
     {
-        public static void Merge<T>(
+        public static Task Merge<T>(
             IEnumerable<ChannelReader<T>> inputs,
             ChannelWriter<T> output,
             CancellationToken stoppingToken)
         {
             if (inputs is null) throw new ArgumentNullException(nameof(inputs));
             if (output is null) throw new ArgumentNullException(nameof(output));
-        
-            async IAsyncEnumerable<T> ReadAllAsync(ChannelReader<T> reader)
-            {
-                if (reader == null) throw new ArgumentNullException(nameof(reader));
-
-                while (await reader.WaitToReadAsync(stoppingToken).ConfigureAwait(false))
-                {
-                    while (reader.TryRead(out T item))
-                    {
-                        yield return item;
-                    }
-                }
-            }
 
-            Task.Run(async () =>
+            return Task.Run(async () =>
                 {
-                    async Task Redirect(ChannelReader<T> input)
-                    {
-                        await foreach (var item in ReadAllAsync(input))
-                            await output.WriteAsync(item, stoppingToken).ConfigureAwait(false);
-                    }
+                    await Task.WhenAll(inputs.Select(input => Redirect(input, output, stoppingToken))
+                            .ToArray())
+                        .ConfigureAwait(false);
 
-                    await Task.WhenAll(inputs.Select(Redirect).ToArray()).ConfigureAwait(false);
                     output.Complete();
```

Comment:
Although the `WhenAll` can only throw when cancelling, it feels like for correctness we should wrap the contents of this method in a try, and do `output.TryComplete(ex);` in the catch block.

Reply:
Yeah I think I agree here, have made this change.